### PR TITLE
[Rebased] fix races

### PR DESCRIPTION
**Pull request by registergoofy**
* proper use of tomb for LeakRoutine
* Fix a bunch of races
* have some locking feature between DumpState and Queue management
* Make sure we empty the overflows when we externally kill LeakRoutine

### CODE REVIEW

File: cmd/crowdsec/crowdsec.go [hunk 1]
```diff
@@ -40,46 +41,75 @@ func initCrowdsec() (*parser.Parsers, error) {
...
-			if err != nil {
-				log.Fatalf("starting parse error : %s", err)
-				return err
+	parserWg := &sync.WaitGroup{}
+	parsersTomb.Go(func() error {
+		parserWg.Add(1)
+		for i := 0; i < cConfig.Crowdsec.ParserRoutinesCount; i++ {
+			parsersTomb.Go(func() error {
+				defer types.CatchPanic("crowdsec/runParse")
+				err := runParse(inputLineChan, inputEventChan, *parsers.Ctx, parsers.Nodes)
```
Comment (buixor) [inline cmd/crowdsec/crowdsec.go:55]: can you inline the error check ?

File: cmd/crowdsec/crowdsec.go [hunk 2]
```diff
@@ -40,46 +41,75 @@ func initCrowdsec() (*parser.Parsers, error) {
...
-				log.Fatalf("starting outputs error : %s", err)
-				return err
-			}
-			return nil
-		})
-	}
+		for i := 0; i < cConfig.Crowdsec.BucketsRoutinesCount; i++ {
+			bucketsTomb.Go(func() error {
+				defer types.CatchPanic("crowdsec/runPour")
+				err := runPour(inputEventChan, holders, buckets, cConfig)
```
Comment (buixor) [inline cmd/crowdsec/crowdsec.go:82]: same

File: cmd/crowdsec/crowdsec.go [hunk 3]
```diff
@@ -40,46 +41,75 @@ func initCrowdsec() (*parser.Parsers, error) {
...
+	})
+	bucketWg.Wait()
+
+	outputWg := &sync.WaitGroup{}
+	outputsTomb.Go(func() error {
+		outputWg.Add(1)
+		for i := 0; i < cConfig.Crowdsec.OutputRoutinesCount; i++ {
+			outputsTomb.Go(func() error {
+				defer types.CatchPanic("crowdsec/runOutput")
+				err := runOutput(inputEventChan, outputEventChan, buckets, *parsers.Povfwctx, parsers.Povfwnodes, *cConfig.API.Client.Credentials)
```
Comment (buixor) [inline cmd/crowdsec/crowdsec.go:101]: same :)

File: cmd/crowdsec/crowdsec.go [hunk 4]
```diff
@@ -90,12 +120,12 @@ func runCrowdsec(parsers *parser.Parsers) error {
...
 }
 
-func serveCrowdsec(parsers *parser.Parsers) {
+func serveCrowdsec(parsers *parser.Parsers, cConfig *csconfig.GlobalConfig) {
 	crowdsecTomb.Go(func() error {
 		defer types.CatchPanic("crowdsec/serveCrowdsec")
 		go func() {
 			defer types.CatchPanic("crowdsec/runCrowdsec")
-			runCrowdsec(parsers)
+			runCrowdsec(cConfig, parsers)
```
Comment (buixor) [inline cmd/crowdsec/crowdsec.go:128]: linter

File: cmd/crowdsec/main.go [hunk 5]
```diff
@@ -200,11 +194,11 @@ func (f *Flags) Parse() {
 }
 
 // LoadConfig return configuration parsed from configuration file
-func LoadConfig(config *csconfig.GlobalConfig) error {
+func LoadConfig(cConfig *csconfig.GlobalConfig) error {
```
Comment (buixor) [inline cmd/crowdsec/main.go:197]: shadow global at https://github.com/crowdsecurity/crowdsec/pull/633/files#diff-ad44e0fba40a2b3f769d325d271d67700077e5c0d6f70ff0b83fade78368ccccR42

File: cmd/crowdsec/serve.go [hunk 6]
```diff
@@ -18,7 +19,7 @@ import (
 )
 
 //debugHandler is kept as a dev convenience : it shuts down and serialize internal state
-func debugHandler(sig os.Signal) error {
+func debugHandler(sig os.Signal, cConfig *csconfig.GlobalConfig) error {
```
Comment (buixor) [inline cmd/crowdsec/serve.go:22]: seems we can kill it ?

File: go.sum [hunk 7]
```diff
@@ -13,6 +13,8 @@ cloud.google.com/go/storage v1.0.0/go.mod h1:IhtSnM/ZTZV8YYJWCY8RULGVqBDmpoyjwiy
 dmitri.shuralyov.com/gpu/mtl v0.0.0-20190408044501-666a987793e9/go.mod h1:H6x//7gZCb22OMCxBHrMx7a5I7Hp++hsVxbQ4BYO7hU=
```
Comment (buixor) [inline go.sum:13]: can you tidy/clean you go.mod / go.sum before commit it ?

**Review by buixor — CHANGES_REQUESTED** [review]
minor changes, lgtm

File: go.sum (hunk 7 above)
Comment (registergoofy) [reply go.sum:13]: done

**Review by buixor — APPROVED** [review]
lgtm

**Pull request merged**